peer/gossip.go

```diff
@@ -0,0 +1,141 @@
+package peer
+
+import (
+	"context"
+	"fmt"
+	"github.com/renproject/aw/dht"
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+	"go.uber.org/zap"
+)
+
+type GossipFunc func(ctx context.Context, subnet id.Hash, contentID []byte) error
+
+// Gossiper returns a new set of callback functions and a gossip function that can be used to spread information throughout a network.
+func Gossiper(logger *zap.Logger, t *transport.Transport, contentResolver dht.ContentResolver, addressTable transport.Table, next Callbacks) (Callbacks, GossipFunc) {
+
+	gossip := func(ctx context.Context, subnet id.Hash, contentID []byte) error {
+		msg := wire.Msg{Version: wire.MsgVersion1, To: subnet, Type: wire.MsgTypePush, Data: contentID}
+
+		var chainedError error = nil
+		var receivers []id.Signatory
+		if subnet.Equal(&GlobalSubnet) {
+			receivers = addressTable.All()
+		} else {
+			receivers = addressTable.Subnet(subnet)
+		}
+		
+		for _, sig := range receivers {
+			addr, ok := addressTable.PeerAddress(sig)
+			if !ok {
+				logger.Error("gossip", zap.String("table", "peer not found"))
+				continue
+			}
+			if err := t.Send(ctx, sig, addr, msg); err != nil {
+				if chainedError == nil {
+					chainedError = fmt.Errorf("%v, gossiping to %v: %v", chainedError, sig, err)
+				} else {
+					chainedError = fmt.Errorf("gossiping to %v: %v", sig, err)
+				}
+			}
+		}
+		return chainedError
+	}
+	didReceivePush := func(from id.Signatory, msg wire.Msg) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		addr, ok := addressTable.PeerAddress(from)
+		if !ok {
+			logger.Error("gossip", zap.Error(ErrPeerNotFound))
+			return
+		}
+
+		hash := [32]byte{}
+		copy(hash[:], msg.Data[:])
+
+		if _, ok := contentResolver.Content(hash); !ok {
```

Comment:
The temptation to make `hash` a byte *array* is because you cannot index a map using a byte *slice*. But, you can index a map using a string, and the conversion `string(byteSlice)` does no allocations and is more-or-less free. Defining the contentID to be `[]byte` in this method (instead of `[32]byte`) will also save you the `copy` from above!